contracts/extensions/CoinMachine.sol

```diff
@@ -130,53 +154,83 @@ contract CoinMachine is ColonyExtension {
 
     targetPerPeriod = _targetPerPeriod;
     maxPerPeriod = _maxPerPeriod;
-    tokensToSell = _tokensToSell;
 
     activePrice = _startingPrice;
     activePeriod = getCurrentPeriod();
 
     emaIntake = wmul(targetPerPeriod, _startingPrice);
 
+    evolvePrice = true;
+
     emit ExtensionInitialised();
   }
 
+  function setWhitelist(address _whitelist) public onlyRoot notDeprecated {
+    whitelist = _whitelist;
+
+    emit WhitelistSet(_whitelist);
+  }
+
   /// @notice Purchase tokens from Coin Machine.
   /// @param _numTokens The number of tokens to purchase
   function buyTokens(uint256 _numTokens) public payable notDeprecated {
     updatePeriod();
 
-    uint256 numTokens = min(min(_numTokens, maxPerPeriod - activeSold), tokensToSell);
+    require(
+      whitelist == address(0x0) || Whitelist(whitelist).isApproved(msg.sender),
+      "coin-machine-unauthorised"
+    );
+
+    uint256 tokenBalance = getTokenBalance();
+    uint256 numTokens = min(min(_numTokens, maxPerPeriod - activeSold), tokenBalance);
     uint256 totalCost = wmul(numTokens, activePrice);
 
     activeIntake = add(activeIntake, totalCost);
     activeSold = add(activeSold, numTokens);
-    tokensToSell = sub(tokensToSell, numTokens);
 
     assert(activeSold <= maxPerPeriod);
 
+    // Check if we've sold out
+    if (numTokens >= tokenBalance) {
+      evolvePrice = false;
```

Comment:
I think we need to emit an event alongside setting this variable, wherever it's set. The front-end needs to be able to not show periods where tokens were unavailable to buy, but show periods where no purchases were made.